File: doordeck-sdk/src/commonMain/kotlin/com/doordeck/multiplatform/sdk/model/LockOperations.kt
package com.doordeck.multiplatform.sdk.model

import com.doordeck.multiplatform.sdk.model.common.UserRole
import kotlinx.datetime.Clock
import kotlin.js.JsExport
import kotlin.jvm.JvmOverloads
import kotlin.time.Duration.Companion.minutes
import kotlin.uuid.Uuid

@JsExport
object LockOperations {

    data class TimeRequirement(
        val start: String, // HH:mm
        val end: String, // HH:mm
        val timezone: String,
        val days: List<String>
    )

    data class LocationRequirement @JvmOverloads constructor(
        val latitude: Double,
        val longitude: Double,
        val enabled: Boolean? = null,
        val radius: Int? = null,
        val accuracy: Int? = null
    )

    data class UnlockBetween @JvmOverloads constructor(
        val start: String, // HH:mm
        val end: String, // HH:mm
        val timezone: String,
        val days: List<String>,
        val exceptions: List<String>? = null
    )

    data class UnlockOperation @JvmOverloads constructor(
        val baseOperation: BaseOperation,
        val directAccessEndpoints: List<String>? = null
    ): Operation

    data class ShareLockOperation(
        val baseOperation: BaseOperation,
        val shareLock: ShareLock
    ): Operation

    data class ShareLock @JvmOverloads constructor(
        val targetUserId: String,
        val targetUserRole: UserRole,
        val targetUserPublicKey: ByteArray,
        val start: Int? = null,
        val end: Int? = null
    )

    data class BatchShareLockOperation(
        val baseOperation: BaseOperation,
        val users: List<ShareLock>
    ): Operation

    data class RevokeAccessToLockOperation(
        val baseOperation: BaseOperation,
        val users: List<String>
    ): Operation

    data class UpdateSecureSettingUnlockDuration(
        val baseOperation: BaseOperation,
        val unlockDuration: Int
    ): Operation

    data class UpdateSecureSettingUnlockBetween @JvmOverloads constructor(
        val baseOperation: BaseOperation,
        val unlockBetween: UnlockBetween? = null
    ): Operation

    data class BaseOperation @JvmOverloads constructor(
        val userId: String? = null,
        val userCertificateChain: List<String>? = null,
        val userPrivateKey: ByteArray? = null,
        val lockId: String,
        val notBefore: Int = Clock.System.now().epochSeconds.toInt(),
        val issuedAt: Int = Clock.System.now().epochSeconds.toInt(),
        val expiresAt: Int = (Clock.System.now() + 1.minutes).epochSeconds.toInt(),
        val jti: String = Uuid.random().toString()
    )

    sealed interface Operation
}

File: doordeck-sdk/src/mingwMain/resources/csharp/Model/LockOperations.cs
namespace Doordeck.Headless.Sdk.Model;

public class TimeRequirement(TimeOnly start, TimeOnly end, TimeZoneInfo timezone, List<DayOfWeek> days)
{
    public TimeOnly Start { get; set; } = start;
    public TimeOnly End { get; set; } = end;
    public TimeZoneInfo Timezone { get; set; } = timezone;
    public List<DayOfWeek> Days { get; set; } = days;
}

public class LocationRequirement(
    double latitude,
    double longitude,
    bool enabled = false,
    int radius = 100,
    int accuracy = 200)
{
    public double Latitude { get; set; } = latitude;
    public double Longitude { get; set; } = longitude;
    public bool Enabled { get; set; } = enabled;
    public int Radius { get; set; } = radius;
    public int Accuracy { get; set; } = accuracy;
}

public class UnlockOperation(BaseOperation baseOperation, List<Uri>? directAccessEndpoints = null)
{
    public BaseOperation BaseOperation { get; set; } = baseOperation;
    public List<Uri>? DirectAccessEndpoints { get; set; } = directAccessEndpoints;
}

public class BaseOperation
{
    public Guid? UserId { get; set; }
    public List<string>? UserCertificateChain { get; set; }
    public string? UserPrivateKey { get; set; }
    public Guid LockId { get; set; }
    public DateTime NotBefore { get; set; }
    public DateTime IssuedAt { get; set; }
    public DateTime ExpiresAt { get; set; }
    public Guid Jti { get; set; }

    public BaseOperation(Guid userId, List<string> userCertificateChain, string userPrivateKey, Guid lockId, DateTime notBefore, DateTime issuedAt, DateTime expiresAt, Guid jti)
    {
        UserId = userId;
        UserCertificateChain = userCertificateChain;
        UserPrivateKey = userPrivateKey;
        LockId = lockId;
        NotBefore = notBefore;
        IssuedAt = issuedAt;
        ExpiresAt = expiresAt;
        Jti = jti;
    }

    public BaseOperation(Guid lockId)
    {
        LockId = lockId;
        NotBefore = DateTime.Now;
        IssuedAt = DateTime.Now;
        ExpiresAt = DateTime.Now.AddMinutes(1);
        Jti = Guid.NewGuid();
    }
}

public class ShareLock(
    Guid targetUserId,
    UserRole targetUserRole,
    byte[] targetUserPublicKey,
    DateTime? start = null,
    DateTime? end = null)
{
    public Guid TargetUserId { get; set; } = targetUserId;
    public UserRole TargetUserRole { get; set; } = targetUserRole;
    public byte[] TargetUserPublicKey { get; set; } = targetUserPublicKey;
    public DateTime? Start { get; set; } = start;
    public DateTime? End { get; set; } = end;
}

public class ShareLockOperation(BaseOperation baseOperation, ShareLock shareLock)
{
    public BaseOperation BaseOperation { get; set; } = baseOperation;
    public ShareLock ShareLock { get; set; } = shareLock;
}

public class BatchShareLockOperation(BaseOperation baseOperation, List<ShareLock> users)
{
    public BaseOperation BaseOperation { get; set; } = baseOperation;
    public List<ShareLock> Users { get; set; } = users;
}

public class RevokeAccessToLockOperation(BaseOperation baseOperation, List<Guid> users)
{
    public BaseOperation BaseOperation { get; set; } = baseOperation;
    public List<Guid> Users { get; set; } = users;
}

public class UpdateSecureSettingUnlockDuration(BaseOperation baseOperation, TimeSpan unlockDuration)
{
    public BaseOperation BaseOperation { get; set; } = baseOperation;
    public TimeSpan UnlockDuration { get; set; } = unlockDuration;
}

public class UnlockBetween(
    TimeOnly start,
    TimeOnly end,
    TimeZoneInfo timezone,
    List<DayOfWeek> days,
    List<DateOnly>? exceptions = null)
{
    public TimeOnly Start { get; set; } = start;
    public TimeOnly End { get; set; } = end;
    public TimeZoneInfo Timezone { get; set; } = timezone;
    public List<DayOfWeek> Days { get; set; } = days;
    public List<DateOnly>? Exceptions { get; set; } = exceptions;
}

public class UpdateSecureSettingUnlockBetween(
    BaseOperation baseOperation,
    UnlockBetween? unlockBetween = null)
{
    public BaseOperation BaseOperation { get; set; } = baseOperation;
    public UnlockBetween? UnlockBetween { get; set; } = unlockBetween;
}